ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdListener.java

```diff
@@ -0,0 +1,107 @@
+package com.applovin.mediation;
+
+import android.content.Context;
+import android.util.Log;
+
+import com.applovin.nativeAds.AppLovinNativeAd;
+import com.applovin.nativeAds.AppLovinNativeAdLoadListener;
+import com.applovin.nativeAds.AppLovinNativeAdPrecacheListener;
+import com.applovin.sdk.AppLovinSdk;
+import com.applovin.sdk.AppLovinSdkUtils;
+import com.google.android.gms.ads.AdRequest;
+import com.google.android.gms.ads.mediation.MediationNativeListener;
+
+import java.lang.ref.WeakReference;
+import java.util.List;
+
+public class AppLovinNativeAdListener
+        implements AppLovinNativeAdLoadListener, AppLovinNativeAdPrecacheListener {
+
+    private AppLovinNativeAdapter adapter;
+    private MediationNativeListener nativeListener;
+    private AppLovinSdk appLovinSdk;
+    private WeakReference<Context> contextWeakReference;
+
+    public AppLovinNativeAdListener(AppLovinNativeAdapter adapter,
+                                    MediationNativeListener nativeListener,
+                                    AppLovinSdk sdk,
+                                    Context context) {
+        this.adapter = adapter;
+        this.nativeListener = nativeListener;
+        this.appLovinSdk = sdk;
+        contextWeakReference = new WeakReference<>(context);
+    }
+
+    @Override
+    public void onNativeAdsLoaded(List<AppLovinNativeAd> nativeAds) {
+        if (nativeAds.size() > 0 && isValidNativeAd(nativeAds.get(0))) {
+            appLovinSdk.getNativeAdService()
+                    .precacheResources(nativeAds.get(0), AppLovinNativeAdListener.this);
```

Comment:
Why not just `this`?

Reply:
We'll never actually return an empty array of native ads, but I like the paranoia! Feel free to keep it in.